### PR TITLE
[BISERVER-11649] Mondrian role mappers don't pass the role name of a user when his username and group are both 'Admin'

### PR DESCRIPTION
**Pull request by IvanNikolaychuk**
**UserRoleDaoResource**

- Track every update of roles, and if this update is touching a user in current session, updating IPentahoSession.SESSION_ROLES attribute, that represents the roles, avalible for current user.

We are doing it via userRoleDaoService, because data there is always up-to-date
- Added few helper methods and changed visibility of some methods for testing reasongs

**UserRoleDaoResource_RoleUpdatesTest**
- A separate test, written specially for checking, that roles were/were not (depending on whether user is session-user or not) updated.

**UserRoleDaoResourceTest**
- Changed few tests a little,(added several mocks), as method content changed.

@rmansoor, @mdamour1976, @pentaho-nbaker, @e-cuellar , review it please

### CODE REVIEW

**Comment by lucboudreau** [comment]
Build Commands:
==============
Build:
------
1. rm -r ~/.kettle; rm -r ~/.pentaho; echo removed hidden folders
2. ant -Djunit.jvmargs=-XX:PermSize=256m -f extensions/build.xml clean-all resolve jacoco jacoco-integration checkstyle publish-local

Cleanup:
-------
rm -r ~/.ivy2/local || echo no publish local to remove

Changed files:
==============
  extensions/src/org/pentaho/platform/web/http/api/resources/UserRoleDaoResource.java
  extensions/test-src/org/pentaho/platform/web/http/api/resources/UserRoleDaoResourceTest.java
  extensions/test-src/org/pentaho/platform/web/http/api/resources/UserRoleDaoResource_RolesUpdatedTest.java


Checkstyle violations:
======================
See: https://github.com/pentaho/pentaho-coding-standards

File extensions/test-src/org/pentaho/platform/web/http/api/resources/UserRoleDaoResourceTest.java
1. Message: Using a static member import should be avoided - org.junit.Assert.assertEquals. at Line: 20 Column: None
2. Message: Using a static member import should be avoided - org.junit.Assert.assertNotNull. at Line: 21 Column: None
3. Message: Using a static member import should be avoided - org.mockito.Matchers.any. at Line: 22 Column: None
4. Message: Using a static member import should be avoided - org.mockito.Matchers.anyString. at Line: 23 Column: None
5. Message: Using a static member import should be avoided - org.mockito.Mockito.*. at Line: 24 Column: None

File extensions/test-src/org/pentaho/platform/web/http/api/resources/UserRoleDaoResource_RolesUpdatedTest.java
1. Message: Using a static member import should be avoided - org.mockito.Mockito.*. at Line: 38 Column: None



Newly Broken Tests:
-----------------
org.pentaho.test.platform.plugin.MetadataQueryComponentIT.testMultiValuedParameters: java.lang.NullPointerException
```
java.lang.NullPointerException
	at org.pentaho.test.platform.plugin.MetadataQueryComponentIT.testMultiValuedParameters(MetadataQueryComponentIT.java:200)

```
org.pentaho.test.platform.plugin.MetadataQueryComponentIT.testPostProcessor: java.lang.NullPointerException
```
java.lang.NullPointerException
	at org.pentaho.test.platform.plugin.MetadataQueryComponentIT.testPostProcessor(MetadataQueryComponentIT.java:337)

```
org.pentaho.test.platform.plugin.MetadataQueryComponentIT.testApplyTemplates: junit.framework.AssertionFailedError
```
junit.framework.AssertionFailedError
	at org.pentaho.test.platform.plugin.MetadataQueryComponentIT.testApplyTemplates(MetadataQueryComponentIT.java:457)

```
org.pentaho.test.platform.plugin.MetadataQueryComponentIT.testJdbcComponent: junit.framework.AssertionFailedError
```
junit.framework.AssertionFailedError
	at org.pentaho.test.platform.plugin.MetadataQueryComponentIT.testJdbcComponent(MetadataQueryComponentIT.java:426)

```
org.pentaho.test.platform.plugin.MetadataQueryComponentIT.testParameters: junit.framework.AssertionFailedError
```
junit.framework.AssertionFailedError
	at org.pentaho.test.platform.plugin.MetadataQueryComponentIT.testParameters(MetadataQueryComponentIT.java:131)

```
org.pentaho.test.platform.plugin.MetadataQueryComponentIT.testEmptyInputs: java.lang.NullPointerException
```
java.lang.NullPointerException
	at org.pentaho.test.platform.plugin.MetadataQueryComponentIT.testEmptyInputs(MetadataQueryComponentIT.java:553)

```
org.pentaho.test.platform.plugin.MetadataQueryComponentIT.testMysqlComponent: junit.framework.AssertionFailedError
```
junit.framework.AssertionFailedError
	at org.pentaho.test.platform.plugin.MetadataQueryComponentIT.testMysqlComponent(MetadataQueryComponentIT.java:383)

```
org.pentaho.test.platform.plugin.MetadataQueryComponentIT.testComponent: junit.framework.AssertionFailedError
```
junit.framework.AssertionFailedError
	at org.pentaho.test.platform.plugin.MetadataQueryComponentIT.testComponent(MetadataQueryComponentIT.java:354)

```
org.pentaho.test.platform.plugin.MetadataQueryComponentIT.testJdbcComponentMaxRows: junit.framework.AssertionFailedError
```
junit.framework.AssertionFailedError
	at org.pentaho.test.platform.plugin.MetadataQueryComponentIT.testJdbcComponentMaxRows(MetadataQueryComponentIT.java:485)

```
org.pentaho.test.platform.plugin.MetadataQueryComponentIT.testPreProcessor: java.lang.NullPointerException
```
java.lang.NullPointerException
	at org.pentaho.test.platform.plugin.MetadataQueryComponentIT.testPreProcessor(MetadataQueryComponentIT.java:297)

```
org.pentaho.test.platform.plugin.KettleIT.testKettleSuccessResult: junit.framework.AssertionFailedError
```
junit.framework.AssertionFailedError: expected:<6> but was:<7>
	at org.pentaho.test.platform.plugin.KettleIT.testKettleSuccessResult(KettleIT.java:68)

```



Still Broken Tests:
-----------------
org.pentaho.test.platform.web.http.api.RepositoryResourceIT.b3_HappyPath_GET_withMimeType: junit.framework.AssertionFailedError
```
junit.framework.AssertionFailedError: Response media type incorrect: expected:<application/pdf;[]charset=UTF-8> but was:<application/pdf;[ ]charset=UTF-8>
	at org.pentaho.test.platform.web.http.api.JerseyTestUtil.assertResponse(JerseyTestUtil.java:47)
	at org.pentaho.test.platform.web.http.api.RepositoryResourceIT.b3_HappyPath_GET_withMimeType(RepositoryResourceIT.java:364)
	at org.springframework.test.context.junit4.statements.RunBeforeTestMethodCallbacks.evaluate(RunBeforeTestMethodCallbacks.java:74)
	at org.springframework.test.context.junit4.statements.RunAfterTestMethodCallbacks.evaluate(RunAfterTestMethodCallbacks.java:83)
	at org.springframework.test.context.junit4.statements.SpringRepeat.evaluate(SpringRepeat.java:72)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.runChild(SpringJUnit4ClassRunner.java:231)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.runChild(SpringJUnit4ClassRunner.java:88)
	at org.springframework.test.context.junit4.statements.RunBeforeTestClassCallbacks.evaluate(RunBeforeTestClassCallbacks.java:61)
	at org.springframework.test.context.junit4.statements.RunAfterTestClassCallbacks.evaluate(RunAfterTestClassCallbacks.java:71)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.run(SpringJUnit4ClassRunner.java:174)

```

Unit test coverage change:
============
org.pentaho.platform.web.http.api.resources.UserRoleDaoResource:
 * instructionPercentCoverageChange : +25.66%
 * branchPercentCoverageChange : +43.38%
 * linePercentCoverageChange : +18.72%
 * complexityPercentCoverageChange : +19.18%
 * methodPercentCoverageChange : +13.48%
    
org.pentaho.platform.plugin.action.kettle.KettleSystemListener:
 * instructionPercentCoverageChange : +1.29%
 * linePercentCoverageChange :**-1.64%**
 * complexityPercentCoverageChange :**-5.88%**
 * methodPercentCoverageChange :**-12.50%**
    
Integration test coverage change:
============
org.pentaho.platform.plugin.services.connections.sql.SQLMetaData:
 * instructionPercentCoverageChange :**-2.80%**
 * linePercentCoverageChange :**-3.33%**
 * complexityPercentCoverageChange :**-9.09%**
 * methodPercentCoverageChange :**-16.67%**
    
org.pentaho.platform.plugin.services.connections.sql.SQLConnection:
 * instructionPercentCoverageChange :**-9.89%**
 * branchPercentCoverageChange :**-14.77%**
 * linePercentCoverageChange :**-9.76%**
 * complexityPercentCoverageChange :**-9.30%**
 * methodPercentCoverageChange :**-9.52%**
    
org.pentaho.platform.plugin.action.pentahometadata.MetadataQueryComponent:
 * instructionPercentCoverageChange :**-44.70%**
 * branchPercentCoverageChange :**-54.35%**
 * linePercentCoverageChange :**-36.97%**
 * complexityPercentCoverageChange :**-23.08%**
 * methodPercentCoverageChange :**-6.25%**
    
org.pentaho.platform.plugin.services.connections.sql.SQLResultSet:
 * instructionPercentCoverageChange :**-0.87%**
 * linePercentCoverageChange :**-1.53%**
 * complexityPercentCoverageChange :**-2.33%**
 * methodPercentCoverageChange :**-5.88%**
    
org.pentaho.platform.plugin.action.kettle.KettleComponent:
 * instructionPercentCoverageChange :**-16.53%**
 * branchPercentCoverageChange :**-12.70%**
 * linePercentCoverageChange :**-23.67%**
 * complexityPercentCoverageChange :**-9.90%**
 * methodPercentCoverageChange :**-45.45%**
    
org.pentaho.platform.plugin.action.chartbeans.ChartAction:
 * instructionPercentCoverageChange : +0.45%

**Comment by IvanNikolaychuk** [comment]
@lucboudreau, I see checkstyle violations, but all of them are static imports of test-library classes, that are based on static imports. 

I also see "newly" falling tests, and drop of code coverage in lots of classes, that I wasn't touching.
Newly falling tests are passing on my computer, and they are **absolutly not related to the changes that I made**, so my changes have no impect on this.
Code coverage on the classes that I changed also increased.

**Comment by lucboudreau** [comment]
The coverage drop is caused by the test failures. A test with an undeterministic outcome is just as bad as a failing one. These failures, although "passing on your machine", should be investigated before we merge this PR. They seem quite simple to investigate, so please spend the time required.

**Comment by IvanNikolaychuk** [comment]
@lucboudreau, you are wrong. I was running both IT, and unit tests, before sending a PR. And thanks, I know, that for running IT-test there is a sepparate command.

**Comment by lucboudreau** [comment]
I did a quick local run and they fail for me too.

**Comment by IvanNikolaychuk** [comment]
@lucboudreau, should I send you a video, where I run this test and it is not failing?

Will take a look at it.

**Comment by lucboudreau** [comment]
Ivan, I do not need a video. I do believe you when you say that the tests pass on your machine. That's not my point. 

We want all tests to pass, everywhere, all of the time. Not just on your machine. Not just here or there. A test which passes on your machine only is absolutely useless to us.

**Comment by IvanNikolaychuk** [comment]
@lucboudreau , sure I do understand it. 

That comment was related to your doubt in my words, but I didn't meen that test should pass everywhere, if it passes on my machine - that would be stupid.

**Comment by IvanNikolaychuk** [comment]
@lucboudreau,  
It can happen because of non-establishing connection with database.
As there were pull-requests after this one, and wingman didn't show this tests falling, I consider, that there was a locale-problem, when running wingman on my PR.

Could you please run it one more time?

**Comment by lucboudreau** [comment]
Build in progress....

**Comment by lucboudreau** [comment]
Build Commands:
==============
Build:
------
1. rm -r ~/.kettle; rm -r ~/.pentaho; echo removed hidden folders
2. ant -Djunit.jvmargs=-XX:PermSize=256m -f extensions/build.xml clean-all resolve jacoco jacoco-integration checkstyle publish-local

Cleanup:
-------
rm -r ~/.ivy2/local || echo no publish local to remove

Changed files:
==============
  extensions/src/org/pentaho/platform/web/http/api/resources/UserRoleDaoResource.java
  extensions/test-src/org/pentaho/platform/web/http/api/resources/UserRoleDaoResourceTest.java
  extensions/test-src/org/pentaho/platform/web/http/api/resources/UserRoleDaoResource_RolesUpdatedTest.java


Checkstyle violations:
======================
See: https://github.com/pentaho/pentaho-coding-standards

File extensions/test-src/org/pentaho/platform/web/http/api/resources/UserRoleDaoResourceTest.java
1. Message: Using a static member import should be avoided - org.junit.Assert.assertEquals. at Line: 20 Column: None
2. Message: Using a static member import should be avoided - org.junit.Assert.assertNotNull. at Line: 21 Column: None
3. Message: Using a static member import should be avoided - org.mockito.Matchers.any. at Line: 22 Column: None
4. Message: Using a static member import should be avoided - org.mockito.Matchers.anyString. at Line: 23 Column: None
5. Message: Using a static member import should be avoided - org.mockito.Mockito.*. at Line: 24 Column: None

File extensions/test-src/org/pentaho/platform/web/http/api/resources/UserRoleDaoResource_RolesUpdatedTest.java
1. Message: Using a static member import should be avoided - org.mockito.Mockito.*. at Line: 38 Column: None







Still Broken Tests:
-----------------
org.pentaho.test.platform.web.http.api.RepositoryResourceIT.b3_HappyPath_GET_withMimeType: junit.framework.AssertionFailedError
```
junit.framework.AssertionFailedError: Response media type incorrect: expected:<application/pdf;[]charset=UTF-8> but was:<application/pdf;[ ]charset=UTF-8>
	at org.pentaho.test.platform.web.http.api.JerseyTestUtil.assertResponse(JerseyTestUtil.java:47)
	at org.pentaho.test.platform.web.http.api.RepositoryResourceIT.b3_HappyPath_GET_withMimeType(RepositoryResourceIT.java:364)
	at org.springframework.test.context.junit4.statements.RunBeforeTestMethodCallbacks.evaluate(RunBeforeTestMethodCallbacks.java:74)
	at org.springframework.test.context.junit4.statements.RunAfterTestMethodCallbacks.evaluate(RunAfterTestMethodCallbacks.java:83)
	at org.springframework.test.context.junit4.statements.SpringRepeat.evaluate(SpringRepeat.java:72)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.runChild(SpringJUnit4ClassRunner.java:231)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.runChild(SpringJUnit4ClassRunner.java:88)
	at org.springframework.test.context.junit4.statements.RunBeforeTestClassCallbacks.evaluate(RunBeforeTestClassCallbacks.java:61)
	at org.springframework.test.context.junit4.statements.RunAfterTestClassCallbacks.evaluate(RunAfterTestClassCallbacks.java:71)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.run(SpringJUnit4ClassRunner.java:174)

```

Unit test coverage change:
============
org.pentaho.platform.web.http.api.resources.UserRoleDaoResource:
 * instructionPercentCoverageChange : +25.66%
 * branchPercentCoverageChange : +43.38%
 * linePercentCoverageChange : +18.72%
 * complexityPercentCoverageChange : +19.18%
 * methodPercentCoverageChange : +13.48%
    
Integration test coverage change:
============
org.pentaho.platform.plugin.action.chartbeans.ChartAction:
 * instructionPercentCoverageChange : +0.45%

**Comment by IvanNikolaychuk** [comment]
@lucboudreau, that tests are passing now, and static imports presents only in tests. Can we merge it now?

**Comment by IvanNikolaychuk** [comment]
@akhayrutdinov, updated

File: extensions/test-src/org/pentaho/platform/web/http/api/resources/UserRoleDaoResourceTest.java [hunk 1]
```diff
@@ -109,6 +107,12 @@ public void testAssignRoleToUser() {
     String user = "testUser1";
     String roles = "testRole1";
 
+    userRoleResource = spy( userRoleResource );
+    IPentahoSession session = mock( IPentahoSession.class );
```
Comment (akhayrutdinov) [inline extensions/test-src/org/pentaho/platform/web/http/api/resources/UserRoleDaoResourceTest.java]: Isn't it possible to use `StandaloneSession` here? I'd prefer to mock as less, as possible

File: extensions/test-src/org/pentaho/platform/web/http/api/resources/UserRoleDaoResourceTest.java [hunk 2]
```diff
@@ -161,6 +165,12 @@ public void testRemoveRolesFromUser() {
...
```
Comment (akhayrutdinov) [inline extensions/test-src/org/pentaho/platform/web/http/api/resources/UserRoleDaoResourceTest.java]: `StandaloneSession` ?

**Comment by akhayrutdinov** [comment]
After this #2810 is merged, it is needed to run WM again. If nothing fails, I'll merge it

**Comment by lucboudreau** [comment]
Build Commands:
==============
Build:
------
1. rm -r ~/.kettle; rm -r ~/.pentaho; echo removed hidden folders
2. ant -Djunit.jvmargs=-XX:PermSize=256m -f extensions/build.xml clean-all resolve jacoco jacoco-integration checkstyle publish-local

Cleanup:
-------
rm -r ~/.ivy2/local || echo no publish local to remove

Changed files:
==============
  extensions/src/org/pentaho/platform/web/http/api/resources/UserRoleDaoResource.java
  extensions/test-src/org/pentaho/platform/web/http/api/resources/UserRoleDaoResourceTest.java
  extensions/test-src/org/pentaho/platform/web/http/api/resources/UserRoleDaoResource_RolesUpdatedTest.java


Checkstyle violations:
======================
See: https://github.com/pentaho/pentaho-coding-standards

File extensions/test-src/org/pentaho/platform/web/http/api/resources/UserRoleDaoResourceTest.java
1. Message: Using a static member import should be avoided - org.junit.Assert.assertEquals. at Line: 20 Column: None
2. Message: Using a static member import should be avoided - org.junit.Assert.assertNotNull. at Line: 21 Column: None
3. Message: Using a static member import should be avoided - org.mockito.Matchers.any. at Line: 22 Column: None
4. Message: Using a static member import should be avoided - org.mockito.Matchers.anyString. at Line: 23 Column: None
5. Message: Using a static member import should be avoided - org.mockito.Mockito.*. at Line: 24 Column: None

File extensions/test-src/org/pentaho/platform/web/http/api/resources/UserRoleDaoResource_RolesUpdatedTest.java
1. Message: Using a static member import should be avoided - junit.framework.Assert.assertTrue. at Line: 40 Column: None
2. Message: Using a static member import should be avoided - org.mockito.Mockito.*. at Line: 41 Column: None









Unit test coverage change:
============
org.pentaho.platform.web.http.api.resources.UserRoleDaoResource:
 * instructionPercentCoverageChange : +25.66%
 * branchPercentCoverageChange : +43.38%
 * linePercentCoverageChange : +18.72%
 * complexityPercentCoverageChange : +19.18%
 * methodPercentCoverageChange : +13.48%
    
org.pentaho.platform.admin.GeneratedContentCleaner:
 * branchPercentCoverageChange : +10.00%
 * complexityPercentCoverageChange : +9.09%
    
Integration test coverage change:
============
org.pentaho.platform.plugin.action.chartbeans.ChartAction:
 * instructionPercentCoverageChange :**-0.45%**

**Comment by IvanNikolaychuk** [comment]
@akhayrutdinov, @pamval  seems like everything is ok now. Could you please merge it?

**Pull request merged**